src/proxy/pool.rs

```diff
@@ -12,80 +12,539 @@
-use hyper_util::client::legacy::pool::{Pool as HyperPool, Poolable, Pooled, Reservation};
-use hyper_util::rt::TokioTimer;
+use std::time::Duration;
+
+use std::collections::hash_map::DefaultHasher;
 use std::future::Future;
+use std::hash::{Hash, Hasher};
 use std::net::IpAddr;
 use std::net::SocketAddr;
-use std::time::Duration;
-use tracing::debug;
+use std::sync::atomic::{AtomicI32, AtomicU16, Ordering};
+use std::sync::Arc;
+
+use tokio::sync::watch;
+use tokio::task;
+
+use tokio::sync::Mutex;
+use tracing::{debug, error, trace};
+
+use crate::config;
+use crate::identity::{Identity, SecretManager};
+
+use flurry::HashMap;
+
+use pingora_pool;
 
-use crate::identity::Identity;
-use crate::proxy::Error;
+// This is merely a counter to track the overall number of conns this pool spawns
+// to ensure we get unique poolkeys-per-new-conn, it is not a limit
+static GLOBAL_CONN_COUNT: AtomicI32 = AtomicI32::new(0);
 
+// A relatively nonstandard HTTP/2 connection pool designed to allow multiplexing proxied workload connections
+// over a (smaller) number of HTTP/2 mTLS tunnels.
+//
+// The following invariants apply to this pool:
+// - Every workload (inpod mode) gets its own connpool.
+// - Every unique src/dest key gets their own dedicated connections inside the pool.
+// - Every unique src/dest key gets 1-n dedicated connections, where N is (currently) unbounded but practically limited
+//   by flow control throttling.
 #[derive(Clone)]
-pub struct Pool {
-    pool: HyperPool<Client, Key>,
+pub struct WorkloadHBONEPool {
+    cfg: config::Config,
+    socket_factory: Arc<dyn SocketFactory + Send + Sync>,
+    cert_manager: Arc<SecretManager>,
+    state: Arc<PoolState>,
+    pool_watcher: watch::Receiver<bool>,
+    max_streamcount: u16,
 }
 
-impl Pool {
-    pub fn new() -> Pool {
-        Self {
-            pool: HyperPool::new(
-                hyper_util::client::legacy::pool::Config {
-                    idle_timeout: Some(Duration::from_secs(90)),
-                    max_idle_per_host: std::usize::MAX,
-                },
-                TokioExec,
-                Some(TokioTimer::new()),
-            ),
+// PoolState is effectively the gnarly inner state stuff that needs thread/task sync, and should be wrapped in a Mutex.
+struct PoolState {
+    pool_notifier: watch::Sender<bool>, // This is already impl clone? rustc complains that it isn't, tho
+    timeout_tx: watch::Sender<bool>, // This is already impl clone? rustc complains that it isn't, tho
+    timeout_rx: watch::Receiver<bool>,
+    // this is effectively just a convenience data type - a rwlocked hashmap with keying and LRU drops
+    // and has no actual hyper/http/connection logic.
+    connected_pool: Arc<pingora_pool::ConnectionPool<ConnClient>>,
+    // this must be an atomic/concurrent-safe list-of-locks, so we can lock per-key, not globally, and avoid holding up all conn attempts
+    established_conn_writelock: HashMap<u64, Option<Arc<Mutex<()>>>>,
```

Comment:
What does? The use of locks?

Reply:
Just locks which don't wrap the thing they're locking...

Reply:
Ah, yep. It's not very cosmetic, agree.

But the kind of locking we have to do is not something the pools out there will do for us, and pingora_pool has _some_ use, at least.

Reply:
Its moot unless we want to fork or contribute to pingora (or I guess lock the entire pool, which we don't care to do) but the point of wrapping the mutex around the thing being accessed is more than cosmetic. It ties holding the lock and accessing the thing together in a way that we can't do with pingora. We'll need to be more careful since we can't do take advantage of that.

Reply:
If we wanted to make this simpler, I would advocate we throw out pingora and implement our own inner data structure from scratch, what we want to do is non-standard and pingora is not likely to want or need what we would have to change to make it work like this - nor is any other "generic" HTTP2 pool. 

Which is why for now I just use it and then do the other stuff outside.